boards/shared/Inc/Io/Io_SharedAdc.h

```diff
@@ -1,67 +1,55 @@
 #pragma once
-
 #include <stm32f3xx_hal.h>
-#include <arm_math.h>
-
-#include "Io_BoardSpecifics.h"
-
-// clang-format off
+#include "Io_SharedAdcConfig.h"
 
-// Memory address where you can find the internal reference voltage calibration 
-// values: Raw ADC data acquired at temperature of 30 degC with VDDA = 3.3V 
-// during the manufacturing process
-#define VREFINT_CALIBRATION_ADDRESS		(uint16_t *)(0x1FFFF7BA)
-
-// Number of microcontroller pins that are configured to be ADC inputs
-#ifndef NUM_ADC_CHANNELS
-    #error "Please define the number pins configured as ADC inputs using NUM_ADC_CHANNELS!"
+#ifndef MAX_NUM_OF_ADC
+#define MAX_NUM_OF_ADC 1
+#endif
+#ifndef MAX_NUM_OF_ADC_CHANNELS
+#define MAX_NUM_OF_ADC_CHANNELS 16
 #endif
 
-/** @brief Nominal voltage for VDDA, or ADC power supply */
-#define VDDA_VOLTAGE (float32_t)(3.3f)
-
-/** @brief Number of VREFINT channel. VREFINT is the internal voltage reference 
- *         that provides a stable (bandgap) voltage output for the ADC. */
-#define NUM_VREFINT_CHANNEL 1
+struct AdcInput;
 
 /**
- * @brief  Initialize ADC in DMA mode and the correct max ADC value
- * @param  hadc ADC handle
- * @param  vrefint_regular_rank The Regular Rank of VREFINT channel configured
- *         in STM32CubeMX (Starts at 1)
+ * Allocate and initialize an ADC instance with direct memory access
+ *
+ * @note The user must ensure that only one of the two ADCs is converting
+ * VREFINT as it is forbidden to have several ADCs converting
+ * VREFINT at the same time. Create the ADC instance containing VREFINT first
+ * while initializing the ADCs
+ *
+ * @note To set up the given ADC with a DMA controller, it must be initialized
+ * with (1) Continuous Conversion Mode enabled,(2) DMA Continuous Requests
+ * enabled, (3) DMA buffer with Circular Mode selected
```

Comment:
nit, but could each numbered item here be on a new line? 
```
 * with:
 *     (1) Continuous Conversion Mode enabled
 *     (2) DMA Continuous Requests enabled
 *     (3) DMA buffer with Circular Mode selected
```

Reply:
@thekenu I think we don't want to use _Continuous Conversion Mode_
It should be triggered periodically by a Timer or launched by software in a periodic task.
I believe @garethellis0 observed that when operating the ADC in _Continuous Conversion Mode_ that the ADC conversions would happen too frequently and it would interfere with other tasks?

Reply:
I super vaguely recall that was a problem but I don't remember all the details now. If the ADC continuous conversion takes up too many CPU cycles then we should consider not using it.

Reply:
Update: I (vaguely) remember the problem was that the ADC callback keeps getting called even though we don't do anything in it. And that's what hogged all the CPU cycle. If that's the case we can maybe investigate a way to disable the callback altogether since we don't do meaningful work in it?

Reply:
By default the callback is a weak function:
[__weak void HAL_ADC_ConvCpltCallback(ADC_HandleTypeDef* hadc)](https://github.com/UBCFormulaElectric/Consolidated-Firmware/blob/master/boards/FSM/Drivers/STM32F3xx_HAL_Driver/Src/stm32f3xx_hal_adc.c#L756)
There's no way to disable the call to that function, it is called after the DMA has transferred the ADC data into memory . We do use this callback. `void HAL_ADC_ConvCpltCallback(ADC_HandleTypeDef* hadc)`  should be implemented in the boards _Io_Adc.c_ file. In `void HAL_ADC_ConvCpltCallback(ADC_HandleTypeDef* hadc)` that's where you want to convert the ADC readings into values or first filter the results and then convert them

Reply:
Are you sure it can't be disabled? My understanding is that the DMA and ADC interrupts are separate. Can you test them and find out?

![image](https://user-images.githubusercontent.com/16970019/81755894-f1aa3a80-946e-11ea-97f3-bf118708573b.png)

As for the converting the ADC readings, this PR seems to handle that in `Io_SharedAdc_GetChannelVoltage` so that's not really a problem. The idea here is to use DMA to store the raw ADC readings, and when the user wants the convert voltage they would call `Io_SharedAdc_GetChannelVoltage` to convert the raw ADC reading.